crates/extensions/c8y_mapper_ext/src/tests.rs

```diff
@@ -2198,6 +2540,244 @@ async fn handle_log_upload_successful_cmd_for_child_device() {
+            "@parent":"device/nested_child//",
+            "@id":"nested_service"
+        })
+        .to_string(),
+    );
+
+    mqtt.send(reg_message).await.unwrap();
+
+    mqtt.send(MqttMessage::new(
+        &Topic::new_unchecked("te/device/nested_child/service/nested_service/a/"),
+        json!({ "severity": "minor", "text": "Temperature high","time":"2023-10-13T15:00:07.172674353Z" }).to_string(),
+    ))
+    .await
+    .unwrap();
+
+    mqtt.skip(3).await;
+
+    // Expect child device service creating minor temperature alarm messages
+    assert_received_contains_str(
+        &mut mqtt,
+        [(
+            "c8y/s/us/nested_service",
+            "303,ThinEdgeAlarm,\"Temperature high\",2023-10-13T15:00:07.172674353Z",
+        )],
+    )
+    .await;
+}
+
+#[tokio::test]
+async fn c8y_mapper_nested_child_service_event_mapping_to_smartrest() {
+    let cfg_dir = TempTedgeDir::new();
+    let (mqtt, _http, _fs, mut timer, _dl) = spawn_c8y_mapper_actor(&cfg_dir, true).await;
+
+    timer.send(Timeout::new(())).await.unwrap(); //Complete sync phase so that alarm mapping starts
+    let mut mqtt = mqtt.with_timeout(TEST_TIMEOUT_MS);
+    skip_init_messages(&mut mqtt).await;
+
+    let reg_message = MqttMessage::new(
+        &Topic::new_unchecked("te/device/immediate_child//"),
+        json!({
+            "@type":"child-device",
+            "@parent":"device/main//",
+            "@id":"immediate_child"
+        })
+        .to_string(),
+    );
+    mqtt.send(reg_message).await.unwrap();
+
+    let reg_message = MqttMessage::new(
+        &Topic::new_unchecked("te/device/nested_child//"),
+        json!({
+            "@type":"child-device",
+            "@parent":"device/immediate_child//",
+            "@id":"nested_child"
+        })
+        .to_string(),
+    );
+    mqtt.send(reg_message).await.unwrap();
+
+    let reg_message = MqttMessage::new(
+        &Topic::new_unchecked("te/device/nested_child/service/nested_service"),
+        json!({
+            "@type":"service",
+            "@parent":"device/nested_child//",
+            "@id":"nested_service"
+        })
+        .to_string(),
+    );
+
+    mqtt.send(reg_message).await.unwrap();
+
+    mqtt.send(MqttMessage::new(
+        &Topic::new_unchecked("te/device/nested_child/service/nested_service/e/"),
+        json!({ "text": "Temperature high","time":"2023-10-13T15:00:07.172674353Z" }).to_string(),
+    ))
+    .await
+    .unwrap();
+
+    mqtt.skip(3).await;
+
+    // Expect nested child device service creating the event messages
+    assert_received_includes_json(
+        &mut mqtt,
+        [(
+            "c8y/event/events/create",
+            json!({
+                "type":"ThinEdgeEvent",
+                "time":"2023-10-13T15:00:07.172674353Z",
+                "text":"Temperature high",
+                "externalSource":{"externalId":"nested_service","type":"c8y_Serial"}
+            }),
+        )],
+    )
+    .await;
+}
+
```

Comment:
Sure, then there were no `assert_messages_matching` APIs, it was tough to assert because of the order issues in the payload. Maybe we can do it as a part of a follow-up task. There are a lot of tests that can be ported.